src/lib/storage/dictionary_segment.cpp

```diff
@@ -1,95 +1,157 @@
 #include "dictionary_segment.hpp"
 
+#include <map>
+#include <set>
+#include "fixed_width_integer_vector.hpp"
+#include "type_cast.hpp"
 #include "utils/assert.hpp"
+#include "value_segment.hpp"
 
 namespace opossum {
 
+std::shared_ptr<AbstractAttributeVector> get_attribute_vector(size_t vector_size, size_t size) {
+  auto bits_needed = std::bit_width(vector_size - 1);
+  Assert(bits_needed <= 32, "Too many values in dictionary, cant use more than 32 bits!");
+  if (bits_needed <= 8) {
+    return std::make_shared<FixedWidthIntegerVector<uint8_t>>(size);
+  } else if (bits_needed <= 16) {
+    return std::make_shared<FixedWidthIntegerVector<uint16_t>>(size);
+  } else {
+    return std::make_shared<FixedWidthIntegerVector<uint32_t>>(size);
+  }
+}
+
 template <typename T>
 DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
-  // Implementation goes here
+  // retrieve value segment
+  const auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(abstract_segment);
+  Assert(value_segment, "Can't encode abstract segment, because it is no value segment.");
+
+  auto values = value_segment->values();
+  auto value_segment_size = value_segment->size();
+  _segment_nullable = value_segment->is_nullable();
+
+  // stores value -> id
+  auto unique_values = std::map<T, ValueID>();
+
+  auto last_index = _segment_nullable ? 1 : 0;
+  for (auto index = size_t{0}; index < value_segment_size; ++index) {
+    if (!value_segment->is_null(index)) {
+      auto value = values[index];
+      if (unique_values.find(value) == unique_values.end()) {
+        unique_values[value] = last_index;
+        ++last_index;
```

Comment:
You could consider inserting the current value using the map insert function 
```suggestion
      const auto inserted = unique_values.insert({value, last_index});
      if (inserted.second) {++last_index;}
```